File: lib/src/utils/traffic_formatter.dart
class TrafficFormatter {
  const TrafficFormatter._();

  static String formatSpeed(int bytesPerSecond) {
    return '${formatBytes(bytesPerSecond)}/s';
  }

  static String formatBytes(int bytes) {
    final value = bytes < 0 ? 0 : bytes;
    const units = ['B', 'KB', 'MB', 'GB', 'TB'];
    var amount = value.toDouble();
    var unitIndex = 0;

    while (amount >= 1024 && unitIndex < units.length - 1) {
      amount /= 1024;
      unitIndex += 1;
    }

    if (unitIndex == 0) {
      return '$value ${units[unitIndex]}';
    }
    if (amount >= 100 || amount == amount.roundToDouble()) {
      return '${amount.round()} ${units[unitIndex]}';
    }
    return '${amount.toStringAsFixed(1)} ${units[unitIndex]}';
  }

  static String formatDuration(Duration duration) {
    final totalSeconds = duration.inSeconds < 0 ? 0 : duration.inSeconds;
    final hours = totalSeconds ~/ 3600;
    final minutes = (totalSeconds % 3600) ~/ 60;
    final seconds = totalSeconds % 60;
    return [
      hours.toString().padLeft(2, '0'),
      minutes.toString().padLeft(2, '0'),
      seconds.toString().padLeft(2, '0'),
    ].join(':');
  }
}

File: plugins/flutter_singbox_vpn/android/src/main/kotlin/com/tecclub/flutter_singbox/utils/TrafficFormatter.kt
package com.tecclub.flutter_singbox.utils

import java.util.Locale
import kotlin.math.roundToInt

object TrafficFormatter {
    private val units = arrayOf("B", "KB", "MB", "GB", "TB")

    fun formatSpeed(bytesPerSecond: Long): String {
        return "${formatBytes(bytesPerSecond)}/s"
    }

    fun formatBytes(bytes: Long): String {
        var amount = bytes.coerceAtLeast(0L).toDouble()
        var index = 0
        while (amount >= 1024.0 && index < units.lastIndex) {
            amount /= 1024.0
            index += 1
        }

        if (index == 0) {
            return "${bytes.coerceAtLeast(0L)} ${units[index]}"
        }

        val value = if (amount >= 100.0 || amount == amount.roundToInt().toDouble()) {
            amount.roundToInt().toString()
        } else {
            String.format(Locale.US, "%.1f", amount)
        }
        return "$value ${units[index]}"
    }

    fun formatDuration(seconds: Long): String {
        val totalSeconds = seconds.coerceAtLeast(0L)
        val hours = totalSeconds / 3600
        val minutes = (totalSeconds % 3600) / 60
        val secs = totalSeconds % 60
        return "%02d:%02d:%02d".format(hours, minutes, secs)
    }
}
